feat: Submission-1

Nightshade14 opened the pull request:
This is the re-submission of submission-1 for the mAIgic team 1.

Team Members:

Satyam Chatrola: sc10247@nyu.edu

Abdullah Suri: aks9171@nyu.edu

Eli Edme: eae8374@nyu.edu

Frank Zhao: bz2190@nyu.edu

Mihir Lovekar: msl9863@nyu.edu

Sidhved Warik: sw6071@nyu.edu

Yaxin Ke: yk2991@nyu.edu

Nightshade14 committed:
feat: Submission-1

Add TrelloManager

feat: Implement OAuth and enhance message handling for Slack bot

fix: hw1 feedback fix

- Upgrade python from 3.11 to 3.12
- Use uv to transform the project from app to a library, exposing api only via modules (modules named with "_" to discourage direct use and import)
- Add mypy for static type checking
- Add pytest-cov for test coverage and reporting
- Add circleCI tests and coverage report
- Untracked artifact files
- The project is capable of being built with build-tools (maigic uses hatchling) and probably being published to PyPi

kyotov commented on .circleci/config.yml:
```diff
@@ -0,0 +1,54 @@
+      - checkout
+
+      - run:
+          name: Get uv via curl and install it
```
depending on how much this time this takes or if you want to fix it at a current version, you may want to build your own image... but this is enough for this project

Nightshade14 replied:
The project is not containerized as of now. And for the CI part, we just need to install dependencies to verify our code standards (with formatting and linting) and functionality (with atleast unit testing).

I chose the image provided by circleCI as it would be most optimized for their systems.

kyotov commented on .circleci/config.yml:
```diff
@@ -0,0 +1,54 @@
+
+      - run:
+          name: Format python files with ruff
+          command: uv run ruff check . --fix
```
ruff also has various checks... 
you probably don't want to run the --fix during CI

Nightshade14 replied:
Would add all the checks that ruff supports and would remove some of the unintuitive ones (by adding to ignore list).

Even though the --fix fixes safe fixes, it might broke code because of any bugs in the ruff package. I would fix it in the next version.

kyotov commented on .python-version:
```diff
@@ -0,0 +1 @@
+3.11
```
why not 3.12?

Nightshade14 replied:
I faced weird mypy issues with python 3.12 and 3.13. Those are tracked issues by mypy and are resolved and would be released in the next mypy version / update.

The mypy error I faced:
`SystemError: extension module 'mypy.nodes' is already cached`

kyotov replied:
let's document the mypy issue and stick with 3.11
but remove this file and put it in `.gitignore`

kyotov commented on README.md:
```diff
@@ -1 +1,47 @@
+# Technical tools
+- Programming Language: Python
+- Project dependency management tool: uv
+- Project linter: ruff
```
i think you want to add mypy also

Nightshade14 replied:
Will add mypy in the next iteration.

kyotov commented:
I tentatively grade this at 95% :)

kyotov commented on LICENSE:
```diff
@@ -1,6 +1,6 @@
 MIT License
 
-Copyright (c) 2024 Kamen Yotov
+Copyright (c) 2024 Satyam Chatrola
```
maybe add more people?

Nightshade14 replied:
Sure.

I will add all the people.

kyotov commented on SlackChatbot/ClientID_Secret.env:
```diff
@@ -0,0 +1,3 @@
+CLIENT_ID = '7855184757495.7892528086800'
+CLIENT_SECRET = '918fec90decd47b23d25fe58e02a73e1'
+BOT_TOKEN='xoxb-7855184757495-7855216149575-kPNmwbhC4mlAyVyUT7QXWpkJ'
```
we need some secret storage for this -- it is bad practice to put in code.

kyotov commented:
i left some comments, but more generally, this is "too much" for HW1.
we need the minimal repository setup -- no extra code for trello, etc.

also a few follow-ups:
* add mypy and ruff and enable all checks -- make sure they run in CI (you have some of this but make it complete)
* add latest python version (i think we are up to 3.13 now?)
* make the repo more "minimal"... trello and friends can come in subsequent PRs
* remove secrets from code.

right now, tentatively this is at about 80/100 for me.
but you can keep improving it.

AbdullahSuri commented:
We have reset the main branch to its initial commit and moved Trello and SlackChatbot enhancements to separate branches. We're working on incorporating the feedback, including adding mypy and ruff checks, updating to Python 3.13, and maintaining a minimal repository setup. Further updates will follow in dedicated PRs.

kyotov commented on src/hello.py:
```diff
@@ -0,0 +1,16 @@
+    return num1 + num2
+
+
+def main() -> int:
+    """Contain the main execution logic."""
+    n1 = 50
+    n2 = 49
+    return addition(n1, n2)
+
+
+if __name__ == "__main__":
+    main()
```
lets have a directory `maigic` under src for the package root
then you can have `api.py` and `_hello.py` inside there.

`api.py`:

```
from ._hello import addition as addition
```

Nightshade14 replied:
Okay. Feedback duly noted

Nightshade14 committed:
fix: hw1 feedback fix
- Upgrade python from 3.11 to 3.12
- Use uv to transform the project from app to a library, exposing api only via modules (modules named with "_" to discourage direct use and import)
- Add mypy for static type checking
- Add pytest-cov for test coverage and reporting
- Add circleCI tests and coverage report
- Untracked artifact files
- The project is capable of being built with build-tools (maigic uses hatchling) and probably being published to PyPi

Nightshade14 commented:
# HW1 fixed

The project is now updated according to the latest feedback. 
## CircleCI tests show up in the tests tab. Samples in case of success and failure are attached.

### CircleCI all tests passed:
<img width="1262" alt="image" src="https://github.com/user-attachments/assets/7c8acdfe-2245-421d-8210-cceb97ec68a9">

### CircleCI all tests failed:
<img width="1280" alt="image" src="https://github.com/user-attachments/assets/ec8fbf50-e039-4dc3-9466-8a3d86312e51">

### CircleCI all tests failed test coverage report:
<img width="1280" alt="image" src="https://github.com/user-attachments/assets/7a308dcd-5749-4d39-ac2f-56b92807d2e0">

## Use this link to check exact diff. There are major changes in the file names and structure, so the diff would be less intuitive.
[Check diff between submissions](https://github.com/nyuoss/mAIgic-nyu/pull/3/commits/40838dd084115d76cee88ff538d006134223d3bd)
